### PR TITLE
Wrap detached torch tensor with dlpack

### PR DESCRIPTION
**Pull request by DavidGinten**
The torch tensor needs to be detached and contiguous
before wrapping it with the dlpack.

### CODE REVIEW

**Comment by DavidGinten** [comment]
* **#8** <a href="https://app.graphite.dev/github/pr/RooflineAI/iree-turbine/8?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/RooflineAI/iree-turbine/8?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `integrate-iree-turbine-20250221`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: iree/turbine/aot/support/ir_utils.py [hunk 1]
```diff
@@ -324,7 +324,8 @@ def create_tensor_global(
                 )
             else:
                 # Emit inline initialized.
-                contents = torch.utils.dlpack.to_dlpack(t)
+                detached_tensor = t.detach().contiguous().cpu()
```
Comment (chrsmcgrr) [inline iree/turbine/aot/support/ir_utils.py]: What is the memory impact here? Because detaching is kind of unclear to me. 

From what I saw in the code, PyTorch as a framework drops ownership of the data and it copies some/all of it somehow. What are the memory impacts here? Can you add a test to our internal suite. 

Could we guard it? Like
```suggestion
                detached_tensor = t
                if not detached_tensor.is_contiguous():
                    detached_tensor = detached_tensor.contiguous()
```
And I don't know about using the `.cpu()` call here. I would expect the model is some kind of FakeTensor/meta device during/after FxGraph export.

**Review by chrsmcgrr — CHANGES_REQUESTED** [review]
Just would like some clarification before we move ahead

File: iree/turbine/aot/support/ir_utils.py (hunk 1 above)
Comment (DavidGinten) [reply iree/turbine/aot/support/ir_utils.py]: That's the line they originally used in turbine. According to [https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html] no copy is performed. And regarding the .cpu(): Can we be sure that the tensor is always on the cpu and the correct device? Because that's what .cpu() does, right? ([https://pytorch.org/docs/stable/generated/torch.Tensor.cpu.html])

Comment (chrsmcgrr) [reply iree/turbine/aot/support/ir_utils.py]: Yeah but I'm not 100% sure the original line is the right thing to do. DLPack is just a view on data. It does not take ownership so detaching semantically does not make sense. 

And for the cpu() case, we know nothing about where we are at this point in the code. I would not change it unless it is strictly required.

Comment (chrsmcgrr) [reply iree/turbine/aot/support/ir_utils.py]: Discussed Offline: this solution seems to give the best result in terms of memory,

File: iree/turbine/aot/support/ir_utils.py [hunk 2]
```diff
@@ -324,7 +324,16 @@ def create_tensor_global(
                 )
             else:
                 # Emit inline initialized.
-                contents = torch.utils.dlpack.to_dlpack(t)
+                # This is the line orignally used here before copying 
+                # the tensor to a memoryview. We also need to make the tensor 
+                # contiguous here to get the coorect strides. We further detach
```
Comment (chrsmcgrr) [inline iree/turbine/aot/support/ir_utils.py]:
```suggestion
                # contiguous here to get the correct strides. We further detach 
```

**Review by chrsmcgrr — APPROVED** [review]

File: iree/turbine/aot/support/ir_utils.py (hunk 1 above)
Comment (DavidGinten) [reply iree/turbine/aot/support/ir_utils.py]: Yeah, running  the pytest has an leak of only 400B in those two lines
